modules/primo/src/Primo/Ting/ValueMapper.php

```diff
@@ -0,0 +1,166 @@
+   *
+   * @param string $langcode
+   *   ISO639 language code.
+   *
+   * @return string
+   *   The mapped language or the input languagecode or the english language-
+   *   name in case of failed mapping.
+   */
+  public static function mapLanguageFromIso639($langcode) {
+    // Languages returned by Primo is in ISO-639 format. To return something
+    // understandable by users we convert it to English and let Drupal try to
+    // translate it to the user language.
+    $lang = (new ISO639())->languageByCode2b($langcode);
+
+    // Return the langcode if we failed to look up the language.
+    if (empty($lang)) {
+      return $langcode;
+    }
+
+    // Do the translation, the english language name will be returned in case
+    // we can't do the translation.
+    return t($lang, [], ['context' => static::T_CONTEXT_LANGUAGE]);
+  }
+
+  /**
+   * Maps translated languages back to a ISO639 language.
+   *
+   * @param string $language
+   *   A language mapped via mapLanguageFromIso639().
+   *
+   * @return string
+   *   The mapped language or original input value in case the mapping fails.
+   */
+  public static function mapLanguageToIso639($language) {
+    // We assume this string came from mapLanguageFromIso639() in which case it
+    // can be a translated language, the english language name or a unmapped
+    // language.
+    // If the original language could not be mapped we might be heading into
+    // trouble. Eg. if the mapping failed and $language is actually a 3-lettered
+    // langcode that happens to be a valid word in the default language, the
+    // attempt to reverse-translate might leave us with something that is far
+    // from a valid langcode.
+    // All of this can happen as T_CONTEXT_LANGUAGE is currently NULL eg. we're
+    // using cores default translations. If we set T_CONTEXT_LANGUAGE to a
+    // custom value we'll avoid all of this, but will have to translate all
+    // possible languages in this context.
+    // Attempt to map the language back to its source.
+    $source = static::reverseTranslate($language, static::T_CONTEXT_LANGUAGE);
+
+    // If this fails we get the input back ($source is now a langcode or a
+    // english language name), if it succeeds we have a english language name.
+    // In both cases it is somewhat safe to attempt to do the mapping back to
+    // ISO639 so go ahead and do it.
+    $langcode = (new ISO639())->code2bByLanguage($source);
+
+    if (empty($langcode)) {
+      // Something went wrong. In case the issue was that $language was actually
+      // a langcode, return it.
+      return $langcode;
+    }
+
+    // Mapping completed, return it.
+    return $langcode;
+  }
+
+  /**
+   * Maps primo genere codes to their mapped counterpart.
+   *
+   * @param string $code
+   *   The Primo code.
+   *
+   * @return string
+   *   The translated code or the input code if it could not be mapped.
+   */
+  public static function mapGenreFromCode($code) {
```

Comment:
This seems separate from the res of the implementation. I think it would be good to have this split into a separate mapper.

Reply:
👍  will consider this for a future refactor.